# Проект по SQL
# Описание проекта
Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать. Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Моя первая задача как аналитика — проанализировать базу данных.

# Цели проекта:
проанализировать базу данных. В базе данных информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

# Задачи

- Посчитать сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитать количество обзоров и среднюю оценку;
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц, чтобы исключить из анализа брошюры;
- Определить автора с самой высокой средней оценкой книг, учитывая только книги с 50 и более оценками;
- Посчитайть среднее количество обзоров от пользователей, которые поставили больше 48 оценок.
# Описание данных


- Таблица books - cодержит данные о книгах:

  - book_id — идентификатор книги;
  - author_id — идентификатор автора;
  - title — название книги;
  - num_pages — количество страниц;
  - publication_date — дата публикации книги;
  - publisher_id — идентификатор издателя.
- Таблица authors - cодержит данные об авторах:

  - author_id — идентификатор автора;
  - author — имя автора.
- Таблица publishers - cодержит данные об издательствах:

  - publisher_id — идентификатор издательства;
  - publisher — название издательства;
- Таблица ratings - cодержит данные о пользовательских оценках книг:

  - rating_id — идентификатор оценки;
  - book_id — идентификатор книги;
  - username — имя пользователя, оставившего оценку;
  - rating — оценка книги.
- Таблица reviews - содержит данные о пользовательских обзорах на книги:

  - review_id — идентификатор обзора;
  - book_id — идентификатор книги;
  - username — имя пользователя, написавшего обзор;
  - text — текст обзора.

In [ ]:
# Импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

import warnings
warnings.filterwarnings('ignore')

# Устанавливаем параметры
db_config = {'user': 'praktikum_student', # Имя пользователя
 'pwd': 'Sdf4$2;d-d30pp', # Пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # Порт подключения
 'db': 'data-analyst-final-project-db'} # Название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

# Сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

# Изучение информации о таблицах

Таблица books cодержит данные о книгах

In [ ]:
query = '''
            SELECT *
            FROM books
        '''
books = pd.io.sql.read_sql(query, con = engine)
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [ ]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


Таблица authors cодержит данные об авторах

In [ ]:
query = '''
            SELECT *
            FROM authors
        '''
authors = pd.io.sql.read_sql(query, con = engine)
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [ ]:
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


Таблица publishers cодержит данные об издательствах

In [ ]:
query = '''
            SELECT *
            FROM publishers
        '''
publishers = pd.io.sql.read_sql(query, con = engine)
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [ ]:
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


Таблица ratings cодержит данные о пользовательских оценках книг

In [ ]:
query = '''
            SELECT *
            FROM ratings
        '''
ratings = pd.io.sql.read_sql(query, con = engine)
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [ ]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


Таблица reviews cодержит данные о пользовательских обзорах на книги

In [ ]:
query = '''
            SELECT *
            FROM reviews
        '''
reviews = pd.io.sql.read_sql(query, con = engine)
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [ ]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


## Задача. Сколько книг вышло после 1 января 2000 года

In [ ]:
books_2000 = '''SELECT COUNT(*)
                FROM books
                WHERE publication_date > '2000-01-01'
                '''

data_books_2000 = pd.io.sql.read_sql(books_2000, con = engine)
print('После 01 января 2020 года вышла {} книга '.format(data_books_2000.loc[0, 'count']))

После 01 января 2020 года вышла 819 книга 


## Задача. Для каждой книги посчитайте количество обзоров и среднюю оценку

In [ ]:
rating = '''SELECT b.title,
            COUNT (DISTINCT rw.review_id), ROUND (avg(r.rating), 2) as R
            FROM REVIEWS AS rw
            RIGHT OUTER JOIN BOOKS AS b ON rw.book_id=b.book_id
            LEFT OUTER JOIN RATINGS AS r ON b.book_id = r.book_id
            GROUP BY b.book_id
            ORDER BY R DESC'''

book_rating = pd.io.sql.read_sql(rating, con = engine)
book_rating

title  count     r
0        Arrows of the Queen (Heralds of Valdemar  #1)      2  5.00
1    The Walking Dead  Book One (The Walking Dead #...      2  5.00
2                                      Light in August      2  5.00
3    Wherever You Go  There You Are: Mindfulness Me...      2  5.00
4    Captivating: Unveiling the Mystery of a Woman'...      2  5.00
..                                                 ...    ...   ...
995  The World Is Flat: A Brief History of the Twen...      3  2.25
996                  His Excellency: George Washington      2  2.00
997                                      Drowning Ruth      3  2.00
998                                              Junky      2  2.00
999                               Harvesting the Heart      2  1.50

[1000 rows x 3 columns]

Посчитано количество обзоров и средняя оценка для каждой книги

## Задача. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц

In [ ]:
books_max = '''SELECT p.publisher,
               COUNT(b.book_id)
               FROM publishers AS p
               JOIN BOOKS AS b ON p.publisher_id=b.publisher_id
               WHERE b.num_pages > 50
               GROUP BY p.publisher_id
               ORDER BY count(b.book_id) DESC
               LIMIT 1'''

max_books = pd.io.sql.read_sql(books_max, con = engine)
max_books

publisher  count
0  Penguin Books     42

Издательство Penguin Books выпустило больше всего книг объемом толще 50 страниц.

## Задача. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [ ]:
autor_max = '''SELECT a.author,
               AVG(r.rating) avg_rating
               FROM authors AS a
               JOIN books AS b ON a.author_id=b.author_id
               JOIN ratings AS r ON b.book_id=r.book_id
               WHERE b.book_id IN
                    (SELECT b.book_id
                     FROM books AS b
                     JOIN ratings AS r ON b.book_id=r.book_id
                     GROUP BY b.book_id
                     HAVING count(r.rating_id)>=50)
               GROUP BY a.author
               ORDER BY avg_rating desc
               LIMIT 1'''

max_autor = pd.io.sql.read_sql(autor_max, con = engine)
max_autor


author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.287097

Наивысшая средняя оценка у J.K. Rowling/Mary GrandPré. Средняя оценка книг - 4.28.

## Задача. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [ ]:
mean_users = '''WITH A AS
                    (SELECT username,
                     COUNT(review_id)
                     FROM reviews
                     WHERE username in
                         (SELECT username
                          FROM ratings
                          GROUP BY username
                          HAVING count(rating_id)>48)
                    GROUP BY username)
                SELECT round (SUM (COUNT) / COUNT(COUNT), 2) AS AVG
                FROM A'''

users_mean = pd.io.sql.read_sql(mean_users, con = engine)


In [ ]:
print ('Среднее количество обзоров от пользователей, которые поставили больше 48 оценок, составляет ', users_mean.loc[0, 'avg'])

Среднее количество обзоров от пользователей, которые поставили больше 48 оценок, составляет  24.0


# Общий вывод
- После 1 января 2000 года было выпущено 819 книг.
- Посчитано количество обзоров и средняя оценка для каждой книги.
- Издательство, выпустившее наибольшее число книг толще 50 страниц: Penguin Books, они выпустили 42 книги.
- Cамая высокая средняя оценка у J.K. Rowling/Mary GrandPré (составляет 4.28).
- Среднее количество обзоров пользователей 24, которые поставили более чем по 48 оценок.